### PR TITLE
Implement Response deserialization

### PR DESCRIPTION
**Pull request by ben-clayton**
Use `Result<>` to handle success / failure results of request handlers. Update the session tests to check for request / response roundtrips, of success / failure messages.

Other changes:
* Remove 'InsertBraces' from .clang-format. Older versions of clang-format can't parse this.
* CMake: export `compile_commands.json`. Consumed by `clangd` and other tools.
* Support `std::array` in `Span`.
* Fix `OneOf operator !=`
* Fix some `std::move`s which should have been `std::forward`.

### CODE REVIEW

File: include/langsvr/session.h [hunk 1]
```diff
@@ -66,24 +71,96 @@ class Session {
...
+        if constexpr (kIsRequest) {
+            return SendRequest(std::forward<T>(message));
+        } else {
+            return SendNotification(std::forward<T>(message));
+        }
+    }
 
+    /// SendRequest encodes and sends the LSP request to the Sender registered with SetSender().
+    /// @param request the request
+    /// @return a Result holding a std::future which will be hold the response value.
```
Comment (amaiorano) [inline include/langsvr/session.h]: s/will be hold/will hold/

File: include/langsvr/traits.h [hunk 2]
```diff
@@ -31,6 +31,12 @@
 #include <string>
 #include <tuple>
 
+/// Forward declaration
+namespace langsvr {
+template <typename SUCCESS_TYPE, typename FAILURE_TYPE>
+struct Result;
+}
```
Comment (amaiorano) [inline include/langsvr/traits.h:38]: Doesn't clang format add `// namespace langsvr` on the closing brace ususally?

**Review by amaiorano — APPROVED** [review]
Good stuff

File: include/langsvr/traits.h (hunk 2 above)
Comment (ben-clayton) [reply include/langsvr/traits.h:38]: I think it omits it if the open and close are close.

File: include/langsvr/session.h (hunk 1 above)
Comment (ben-clayton) [reply include/langsvr/session.h]: Done

**Pull request merged**